src/Microsoft.Extensions.Http.Polly/DependencyInjection/PollyHttpClientBuilderExtensions.cs

```diff
@@ -74,6 +75,96 @@ public static IHttpClientBuilder AddPolicyHandler(this IHttpClientBuilder builde
             return builder;
         }
 
+        public static IHttpClientBuilder AddPolicyHandler(
+            this IHttpClientBuilder builder, 
+            Func<HttpRequestMessage, IAsyncPolicy<HttpResponseMessage>> policyFactory)
+        {
+            if (builder == null)
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            if (policyFactory == null)
+            {
+                throw new ArgumentNullException(nameof(policyFactory));
+            }
+
+            builder.AddHttpMessageHandler(() => new PolicyHttpMessageHandler(policyFactory));
+            return builder;
+        }
+
+        public static IHttpClientBuilder AddPolicyHandler(
+            this IHttpClientBuilder builder,
+            Func<IServiceProvider, HttpRequestMessage, IAsyncPolicy<HttpResponseMessage>> policyFactory)
+        {
+            if (builder == null)
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            if (policyFactory == null)
+            {
+                throw new ArgumentNullException(nameof(policyFactory));
+            }
+
+            builder.AddHttpMessageHandler((services) =>
+            {
+                return new PolicyHttpMessageHandler((request) => policyFactory(services, request));
+            });
+            return builder;
+        }
+
+        public static IHttpClientBuilder AddPolicyHandlerFromRegistry(this IHttpClientBuilder builder, string policyKey)
+        {
+            if (builder == null)
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            if (policyKey == null)
+            {
+                throw new ArgumentNullException(nameof(policyKey));
+            }
+
+            builder.AddHttpMessageHandler((services) =>
+            {
+                var registry = services.GetRequiredService<IReadOnlyPolicyRegistry<string>>();
+
+                // The policy might be either IAsyncPolicy or IAsyncPolicy<HttpRequestMessage> but never both.
+                // But, the handler expects IAsyncPolicy<HttpRequestMessage>, so try first for what we want and then
+                // fall back to the non-generic interface. We'll get the non-generic case throw the policy isn't registred.
+                if (!registry.TryGet<IAsyncPolicy<HttpResponseMessage>>(policyKey, out var policy))
+                {
+                    policy = registry.Get<IAsyncPolicy>(policyKey).WrapAsync(Policy.NoOpAsync<HttpResponseMessage>());
+                }
```

Comment:
@reisenberger - there's a lot of tension caused in this design by the desire to support `IAsyncPolicy` and `IAsyncPolicy<HttpResponseMessage>`.

For instance if we were to support `Func<HttpRequestMessage, IAsyncPolicy<HttpResponseMessage>>` **and** `Func<HttpRequestMessage, IAsyncPolicy>`, then we end up with a method that's overload on the return type of `Func<..T>`. In general this case gives really bad intellisense and I try to avoid it.

It seems like it's always possible to create a `PolicyAsync<T>` instead of a `PolicyAsync` for anything you desire. Thoughts?

Reply:
>a lot of tension caused in this design by the desire to support `IAsyncPolicy` and `IAsyncPolicy<HttpResponseMessage>`.

Supporting only `IAsyncPolicy<HttpResponseMessage>` would certainly simplify the implementation.  With `HttpClientFactory` there a better case for supporting only generic policies than with Polly in general, given all calls are in `HttpResponseMessage`.

A concern is that not supporting non-generic `IAsyncPolicy` will lead to enquiries of why, eg, `.AddPolicyHandler(Policy.TimeoutAsync(10))` doesn't compile. In Polly's API, I opted to take the hit and provide bridging overloads ([eg in PolicyWrap](https://github.com/App-vNext/Polly/blob/754a4a920900e1ae56d4fdae7cbdfd02bcb8cbd1/src/Polly.Shared/Wrap/PolicyWrapSyntaxAsync.cs#L44-L51)), so that from the user's perspective, "everything just worked".  'Traffic-managing' resilience strategies like Timeout and Bulkhead have no interest in returned results, so we opted to ensure that non-generic forms could [just be used](https://github.com/App-vNext/Polly/wiki/Non-generic-and-generic-policies#mixing-non-generic-and-generic-in-a-policywraptresult) wherever a generic form might otherwise be expected.

If the decision is that `HttpClientFactory` only supports `IAsyncPolicy<HttpResponseMessage>`, documentation should cover this, to reduce user blocks, and enquiries.

---

Polly could add a helper extension method `Policy<TResult> Policy.For<TResult>()` (the equivalent of your `.WrapAsync(Policy.NoOpAsync<HttpResponseMessage>())` without the need for the extra `NoOpPolicy<TResult>`).  This could either help users bridge; or `HttpClientFactory` provide the extra overloads.

---

Worth further thought and discussion, as will be key to the user experience.

Reply:
I had a discussion this morning about this with @glennc - we're thinking that we want to try removing/not-providing the non-generic methods for the preview release and see if we get feedback.

There's a relatively small delta between `.DoThing(Policy.Timeout(...))` and `.DoThing(Policy.Timeout<HttpResponseMessage>())`. We'd rather require the extra typing than ship partial or complicated support for non-generic policies.

If we get the feedback that this is important then we will do the work to add overloads that directly accept non-generic `IAsyncPolicy`.

--

I agree that it would be a nice addition to have a simpler way to go from `IAsyncPolicy`->`IAsyncPolicy<>`. I suppose that's on you guys if you think it's worth doing. For now I plan to just document how to do it.

Reply:
I think this is a balanced approach

We don't really know whether people will miss the non-generic methods so it's probably fine to leave it out and see who complains.

Reply:
@rynowak Very happy with that call :+1:; I agonised over this. 

Offering the mix of non-generic and generic overloads makes sense for Polly generally, as we have a range of use cases to support, some of which suit non-generic policies better; but with `HttpClientFactory`'s one use case (for Polly), there is a chance to reduce complexity.  Simplicity is a great virtue.